---
title: "Prediction of analysis success using sequencing quality metrics"
author: "Jouke Profijt"
date: "2/4/2020"
output: pdf_document
---

# QC reports
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{sh eval=FALSE}
ls !(*pdf) | grep -P 'sampleidMatcher'
```
returns produced metrics

* alignment_summary_metrics
* bam_index_stats
* flagstat
* gc_bias_metrics
* hs_metrics
* insert_size_metrics
* nonAutosomalRegionChrX_hs_metrics
* quality_by_cycle_metrics
* quality_distribution_metrics


## alignment summary metrics

the Picard alignment summary metrics application is a quality control tool for exome-Seq and RNA-seq aligments. it provides lots of metrics.

### CATEGORY
Describes the type of read.

either `UNPAIRED`, `FIRST_OF_PAIR`, `SECOND_OF_PAIR` or `PAIR`. Unpaired means the read is part of a fragmented run. First of pair are reads for only the first read in a paired run, and subsuquently Second of pair are only the second read in a paired run. Pair means the metrics are aggregated for both the first and second reads in a pair.

### PF_READS
This statistic defines the number of PF reads, where PF means the reads that have passed the illumina filter.

### TOTAL_READS
Total number of reads including all PF and non-PF reads

### PCT_PF_READS
The fraction of the reads that have passed the filter/PF reads. `PF_READS/TOTAL_READS`

### PF_NOISE_READS
The number of PF reads that are marked as noise. These reads are reads entirely composed of A bases and or N bases.

These reads are usally artificially created.

### PF_READS_ALIGNED
The number of PF reads that were aligned to the reference sequence

### PCT_PF_READS_ALIGNED
The percentage of PF reads aligned to the reference sequence

### PF_ALIGNED_BASES
the total number of aligned bases

### PF_HQ_ALIGNED_READS
The number of PF reads that were aligned to the reference sequence with a mapping quality of Q20 or higher signifying that the aligner estimates a 1/100 chance that the alignment is wrong.

### PF_HQ_ALIGNED_BASES
the number of bases aligned to the reference sequence in reads that were mapped at a high qualiy. usally aproxomatly `PF_HQ_ALIGNED_READS*READ_LENGHT`

### PF_HQ_ALIGNED_Q20_BASES
a Subset of `PF_HQ_ALIGNED_BASES` where the base call quality was !20 or higher

### PF_HQ_NEDUAB_MISMATCHES
The median number of mismatches versus the reference sequence in reads that were aligned to the reference at high quality

### PF_MISMATCH_RATE
the rate of bases mismatching the refference for all bases aligned to the reference sequence

### PF_HQ_ERROR_RATE
The fraction of bases that mismatch the reference in PF HQ aligned reads

### PF_INDEL_RATE
the number of insertion and deletion events per 100 aligned bases. Uses the number of events as the numerator, not the number of inserted or deleted bases

### MEAN_READ_LENGTH
the mean read lenght of the set of reads examined. When looking at the data for a single lane with equal lenght reads this number is just the read lenght. When looking at data for the merged lanes with differeing read lenghts this is the mean read lenght of all bases.

### READS_ALIGNED_IN_PAIRS
The number of reads whose mate pair was also aligned to the reference

### PCT_READS_ALIGNED_IN_PAIRS
The fraction of reads whose mate pair was also aligned to the reference

### PF_READS_IMPROPER_PAIRS
the number of (primary) aligned reads that are **not** "properly" aligned in pairs

### PCT_PF_READS_IMPROPER_PAIRS
The fraction of (primary) reads that are *not* "properly" aligned in pairs (as per SAM flag 0x2). PF_READS_IMPROPER_PAIRS / PF_READS_ALIGNED

### BAD_CYCLES
The number of instrument cycles in which 80% or more of base calls were no-calls

### STRAND_BALANCE
The number of PF reads aligned to the positive strand of the genome divided by the number of PF reads aligned to the genome.

### PCT_CHIMERAS
The fraction of reads that map outside of a maximum insert size (usually 100kb) or that have the two ends mapping to different chromosomes.

### PCT_ADAPTER
The fraction of PF reads that are unaligned and match to a known adapter sequence right from the start of the read.


## Bam index stats

includes the counts of aligned and unalligned reads, and reads with no start coordinates.

## flagstat
[samtools-flagstat](https://www.htslib.org/doc/samtools-flagstat.html)

Provides counts for each of 13 categories based primarily on bit flags in the FLAG field. Each category is broken up into QC pass or QC fail. For example the first row in the flagstat file will look like:

122 + 28 in total (QC-passed reads + QC-failed reads)

which means that of the total 150 reads 122 were marked as QC pass and 28 as QC fail. The other categories are in the same structure

## gc bias metrics
[picard metric definitions](https://personal.broadinstitute.org/picard/picard_metric_definitions.html)
Metrics class: GcBiasDetailMetrics: Class that holds detailed metrics about reads that fall within windows of a certain GC bin on the reference genome.

GC: The G+C content of the reference sequence represented by this bin. Values are from 0% to 100%
WINDOWS: The number of windows on the reference genome that have this G+C content.
READ_STARTS: The number of reads whose start position is at the start of a window of this GC.
MEAN_BASE_QUALITY: The mean quality (determined via the error rate) of all bases of all reads that are assigned to windows of this GC.
NORMALIZED_COVERAGE: The ration of "coverage" in this GC bin vs. the mean coverage of all GC bins. A number of 1 represents mean coverage, a number less than one represents lower than mean coverage (e.g. 0.5 means half as much coverage as average) while a number greater than one represents higher than mean coverage (e.g. 3.1 means this GC bin has 3.1 times more reads per window than average).
ERROR_BAR_WIDTH: The radius of error bars in this bin based on the number of observations made. For example if the normalized coverage is 0.75 and the error bar width is 0.1 then the error bars would be drawn from 0.65 to 0.85.

## hs_metrics
[hs_metrics-docs](https://personal.broadinstitute.org/picard/picard_metric_definitions.html#HsMetrics)

## insert size metrics
Metrics about the insert size distribution of a paired-end library, created by the CollectInsertSizeMetrics program and usually written to a file with the extension ".insert_size_metrics". In addition the insert size distribution is plotted to a file with the extension ".insert_size_Histogram.pdf".

[insert size metrics docs](https://personal.broadinstitute.org/picard/picard_metric_definitions.html#InsertSizeMetrics)

## nonAutosomalRegionChrX_hs_metrics
hs metrics for Chromosome X

## quality_by_cycle_metrics

This metric gives an overall snapshot of sequencing machine performance. For most types of sequencing data, the output is expected to show a slight reduction in overall base quality scores towards the end of each read. Spikes in quality within reads are not expected and may indicate that technical problems occurred during sequencing

## quality distribution metrics
This tool is used for determining the overall 'quality' for a library in a given run. To that effect, it outputs a chart and tables indicating the range of quality scores and the total numbers of bases corresponding to those scores. Options include plotting the distribution of all of the reads, only the aligned reads, or reads that have passed the Illumina Chastity filter thresholds as described here.

## files
```{r}
prm05 <- read.csv("data/prm05-projects-by-time", header = FALSE, sep = "")
colnames(prm05) <- c("permissions", "blocks", "user", "group", "size", "month", "day", "time", "name")
```

# InterOp
The interop files are binary files produced during a run providing detailed statistics. They are located in the rawdata of a run `run/Info/InterOp`.


## CorrectedIntMetricsOut.bin
Per tile per cycle per channel average intensity values

## ExtractionMetricsOut.bin
Per tile per cycle per channel 90th percentile of intensity and FWHM values

## EmpiricalPhasingMetricsOut.bin
Phasing weights per tile per cycle

## QMetricsOut.bin
Per tile per cycle subsampled basecall quality score histogram

## TileMetricsOut.bin
Per tile information about density, cluster count, passed filter, percent aligned

## PFGridMetricsOut.bin
Not publically documented

## EventMetricsOut.bin
Not publically documented

## RegistrationMetricsOut.bin 
Not publically documented

