---
title: "Correlations"
author: "Jouke Profijt"
date: "4/9/2020"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = 'C:/Users/Jouke/Documents/analysis-metrics-prediction-display' )
```

```{r echo=FALSE}
library(DBI)
library(knitr)
library(ggplot2)

## establish connection to sqlite db
connection = dbConnect(RSQLite::SQLite(), "data/SQLITE/databaseFull.db")

res <- dbSendQuery(connection, "
                  SELECT
                  Samples.ID as SampleID,
                  RUNS.Date,
                  Samples.Sequencer,
                  Samples.capturingKit as Kit,
                  avg(RunSummary.Yield) as Yield,
                  avg(RunSummary.ProjectedYield) as ProjectedYield,
                  avg(RunSummary.Intensity) as Intensity,
                  avg(RunSummary.Q30) as RunQ30,
                  avg(Lanes.Tiles) as Tiles,
                  avg(Lanes.DensityMIN) as DensitySD,
                  avg(Lanes.DensityMAX) as ClusterDensity,
                  avg(Lanes.ClusterMIN) as ClusterSD,
                  avg(Lanes.ClusterMAX) as ClustersPCT,
                  avg(Lanes.LegacyPhasing) as LegacyPhasing,
                  avg(Lanes.LegacyPrePhasing) as LegacyPrePhasing,
                  avg(Lanes.PhasingSlope) as PhasingSlope,
                  avg(Lanes.PrePhasingSlope) as PrePhasingSlope,
                  avg(Lanes.PhasingOffset) as PhasingOffset,
                  avg(Lanes.PrePhasingOffset) as PrePhasingOffset,
                  AlignmentSummaryMetrics.PFHQmedianMismatches,
                  AlignmentSummaryMetrics.PFmismatchRate,
                  AlignmentSummaryMetrics.PFHQErrorRate, 
                  AlignmentSummaryMetrics.PFindelRate,
                  AlignmentSummaryMetrics.MeanReadLenght,
                  AlignmentSummaryMetrics.ReadsAllignedInPairs,
                  AlignmentSummaryMetrics.BadCycles,
                  AlignmentSummaryMetrics.StrandBalance,
                  AlignmentSummaryMetrics.ChimerasPercentage,
                  AlignmentSummaryMetrics.AdapterPercentage,
                  InsertSizes.MedianSize,
                  InsertSizes.MedianAbsoluteDeviation,
                  InsertSizes.MinSize,
                  InsertSizes.MaxSize,
                  InsertSizes.MeanSize,
                  InsertSizes.StandardDeviation,
                  InsertSizes.ReadPairs,
                  InsertSizes.W10,
                  InsertSizes.W20,
                  InsertSizes.W30,
                  InsertSizes.W40,
                  InsertSizes.W50,
                  InsertSizes.W60,
                  InsertSizes.W70,
                  InsertSizes.W80,
                  InsertSizes.W90,
                  InsertSizes.W99,
                  hsMetrics.TotalReads,
                  hsMetrics.BaitDesignEfficientcy,
                  hsMetrics.PFreads,
                  (hsMetrics.PFreads - hsMetrics.PFuniqueReads) as DuplicateReads,
                  hsMetrics.PFuniqueReads,
                  hsMetrics.PFUQaligned,
                  (hsMetrics.PFBasesAligned/1000000) as PFMegaBasesAligned,
                  (hsMetrics.PFUQBasesAligned/1000000) as PFUQMegaBasesAligned,
                  (hsMetrics.OnBaitBases/1000000) as OnBaitMegaBases,
                  (hsMetrics.NearBaitBases/1000000) as NearBaitMegaBases ,
                  (hsMetrics.OffBaitBases/1000000) as OffBaitMegaBases,
                  hsMetrics.MeanBaitCoverage,
                  hsMetrics.MeanTargetCoverage,
                  hsMetrics.MedianTargetCoverage,
                  hsMetrics.PercentageUsableBasesOnBait,
                  hsMetrics.PercentageUsableBasesOnTarget,
                  hsMetrics.FoldEnrichment,
                  hsMetrics.TargetBasesPct20X,
                  FlagstatMetrics.TotalPass,
                  FlagstatMetrics.SecondaryPass,
                  FlagstatMetrics.SupplementaryPass,
                  FlagstatMetrics.DuplicatePass,
                  FlagstatMetrics.MappedPass,
                  FlagstatMetrics.MappedPercentage,
                  FlagstatMetrics.PairedSeqPass,
                  FlagstatMetrics.Read1Pass,
                  FlagstatMetrics.Read2Pass,
                  FlagstatMetrics.PoperPairPass,
                  FlagstatMetrics.ProperPairPCT,
                  FlagstatMetrics.SelfAndMatePass,
                  FlagstatMetrics.SingletonsPass,
                  FlagstatMetrics.SingletonsPercentage,
                  FlagstatMetrics.MateOnDiffChromosomeLowPass,
                  FlagstatMetrics.MateOnDiffChromosomeHighPass
                  FROM Samples 
                  INNER JOIN RUNS ON Samples.Sequencer==RUNS.Sequencer AND Samples.Run==RUNS.Number AND Samples.startDate==RUNS.Date 
                  INNER JOIN RunSummary ON RUNS.UniqueID==RunSummary.UniqueID
                  INNER JOIN Lanes ON RUNS.UniqueID==Lanes.UniqueID
                  LEFT JOIN AlignmentSummaryMetrics ON Samples.ID==AlignmentSummaryMetrics.SampleID
                  INNER JOIN InsertSizes ON Samples.ID==InsertSizes.SampleID
                  INNER JOIN hsMetrics ON Samples.ID==hsMetrics.SampleID
                  INNER JOIN FlagstatMetrics ON Samples.ID==FlagstatMetrics.SampleID
                  WHERE AlignmentSummaryMetrics.Category=='PAIR' AND Lanes.ReadID NOT LIKE '%(I)'
                  GROUP BY Samples.ID
                  ORDER BY RUNS.Date")
Metrics.per.Sample <- dbFetch(res)
dbClearResult(res)
Metrics.per.Sample.pl <- Metrics.per.Sample[,-1]
rownames(Metrics.per.Sample.pl) <-Metrics.per.Sample[,1]
Metrics.per.Sample <- Metrics.per.Sample.pl
rm(Metrics.per.Sample.pl)

samples <- dbReadTable(connection, "Samples")
dbDisconnect(connection)
rm(connection, res)

## Format the dates
Metrics.per.Sample$Date <- as.Date(Metrics.per.Sample$Date)
Metrics.per.Sample$Date <- as.factor(weekdays(Metrics.per.Sample$Date))
colnames(Metrics.per.Sample)[1] <- "Weekday"
Metrics.per.Sample$Sequencer <- as.factor(Metrics.per.Sample$Sequencer)
Metrics.per.Sample$Kit <- as.factor(Metrics.per.Sample$Kit)
Interop.per.Sample <- Metrics.per.Sample[4:18]
dependend <-  Metrics.per.Sample[19:79]
ASM.per.Sample <- Metrics.per.Sample[19:28]
IS.per.Sample <- Metrics.per.Sample[29:45]
hs.per.Sample <- Metrics.per.Sample[46:63]
fl.per.Sample <- Metrics.per.Sample[64:79]
samples$DNA_numbers <- as.factor(apply(samples, 1, function(x) {
  ids <- unlist(strsplit(x[1], "_"))
  ids[3]
  }))

samples$DesignNumber <- as.factor(apply(samples, 1, function(x) {
  ids <- unlist(strsplit(x[1], "_"))
  ids[6]
  }))

samples <- subset(samples, ((grepl("DNA\\d+", DNA_numbers) )))
head(samples)
levels(samples$UniqueKits)
```

```{r}
library(DBI)
library(ggpubr)
library(gridExtra)
connection <- dbConnect(RSQLite::SQLite(), "data/SQLITE/databaseFull.db")
res <- dbSendQuery(connection, "
                  SELECT
                  RUNS.Date,
                  avg(RunSummary.Yield) as Yield,
                  avg(RunSummary.ProjectedYield) as ProjectedYield,
                  avg(RunSummary.Intensity) as Intensity,
                  avg(RunSummary.Q30) as RunQ30,
                  avg(Lanes.Tiles) as Tiles,
                  avg(Lanes.DensityMAX) as ClusterDensity,
                  avg(Lanes.ClusterMAX) as ClustersPCT,
                  avg(Lanes.LegacyPhasing) as LegacyPhasing,
                  avg(Lanes.LegacyPrePhasing) as LegacyPrePhasing,
                  avg(Lanes.PhasingSlope) as PhasingSlope,
                  avg(Lanes.PrePhasingSlope) as PrePhasingSlope,
                  avg(Lanes.PhasingOffset) as PhasingOffset,
                  avg(Lanes.PrePhasingOffset) as PrePhasingOffset,
                  avg(Lanes.Reads) as Reads,
                  avg(Lanes.ReadsPF) as ReadsPF,
                  avg(Lanes.Q30) as LaneQ30,
                  avg(Lanes.IntensityMIN) as IntensityMIN,
                  avg(Lanes.IntensityMAX) as IntensityMAX
                  FROM RUNS
                  INNER JOIN RunSummary ON RUNS.UniqueID==RunSummary.UniqueID
                  INNER JOIN Lanes ON RUNS.UniqueID==Lanes.UniqueID
                  WHERE Lanes.ReadID NOT LIKE '%(I)'
                  GROUP BY RUNS.UniqueID
                  ORDER BY RUNS.Date")

interopFull <- dbFetch(res)
dbClearResult(res)

pRq30 <- ggdensity(interopFull$RunQ30, 
          main = "Density plot of Percentage of q30 reads per run",
          xlab = "Q30 percentage")
pIn <- ggdensity(interopFull$Intensity, 
          main = "Density plot of c1 Intensity",
          xlab = "c1 Intensity")
pRq30 <- ggdensity(interopFull$RunQ30, 
          main = "Density plot of Percentage of q30 reads per run",
          xlab = "Q30 percentage")
pIn <- ggdensity(interopFull$Intensity, 
          main = "Density plot of c1 Intensity",
          xlab = "c1 Intensity")
pCl <- ggdensity(interopFull$ClustersPCT, 
          main = "Density plot of clusters passing filter",
          xlab = "clustes passing filter (%)")
pD <- ggdensity(interopFull$ClusterDensity, 
          main = "Density plot of cluster density",
          xlab = "Cluster dencity (K/mm2)")
pPh <- ggdensity(interopFull$LegacyPhasing, 
          main = "Density plot of phasing",
          xlab = "Phasing (%)")
pPre <- ggdensity(interopFull$LegacyPrePhasing, 
          main = "Density plot of PrePhasing",
          xlab = "PrePhasing (%)")
grid.arrange(pRq30, pIn, pCl, pD, pPh, pPre, nrow = 3)
```
Most Interop metrics are not normally distributed - pearson assumes normally distrubuted variables,  therefore to compute correlations the spearman correlation method will be used.

```{r}
library(ggplot2)
library(reshape2)
library(dplyr)
dup.interop <- duplicated(Interop.per.Sample)
unique.interop <- Interop.per.Sample[!dup.interop,]
unique.asm <- ASM.per.Sample[!dup.interop,]

corMat <- round(cor(x=unique.interop, y=unique.asm, method = "spearman", use = "complete.obs"), 2)



ASM.cor.heatmap <- ggplot(data = melt(corMat, na.rm = T), aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile(colour = "white") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                       midpoint = 0, limit = c(-1,1), space = "Lab", 
                       name="Spearman\nCorrelation") + 
  theme_minimal()+ theme(legend.position = "none") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                   size = 12, hjust = 1)) + xlab("Interop Metrics") + ylab("Alignment Summary Metrics") + ggtitle("Interop vs Alignemnt Summary Metrics")

ASM.cor.heatmap
ggsave(
  "asm.png",
  plot = ASM.cor.heatmap,
  dpi = 300,
  limitsize = TRUE
)
```

```{r}
library(ggplot2)
library(reshape2)
library(gridExtra)


corMat <- round(cor(x=Interop.per.Sample, y=IS.per.Sample[1:7], method = "spearman", use = "complete.obs"), 2)



IS.cor.heatmap <- ggplot(data = melt(corMat, na.rm = T), aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile(colour = "white") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                       midpoint = 0, limit = c(-1,1), space = "Lab", 
                       name="Spearman\nCorrelation") + 
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                   size = 12, hjust = 1)) + xlab("Interop Metrics") + ylab("Insert Size Metrics") + theme(legend.position = "none") + ggtitle("Interop vs InsertSize metrics")

corMat <- round(cor(x=Interop.per.Sample, y=IS.per.Sample[8:17], method = "spearman", use = "complete.obs"), 2)
IS.widths.cor.heatmap <- ggplot(data = melt(corMat, na.rm = T), aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile(colour = "white") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                       midpoint = 0, limit = c(-1,1), space = "Lab", 
                       name="Spearman\nCorrelation") + 
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                   size = 12, hjust = 1)) + xlab("Interop Metrics") + theme(axis.title.y = element_blank())

 is <- grid.arrange(IS.cor.heatmap, IS.widths.cor.heatmap, nrow=1)
 
 ggsave(
  "insertsizes.png",
  plot = IS.cor.heatmap,
  dpi = 300,
  limitsize = T)
```

```{r}
library(ggplot2)
library(reshape2)
library(dplyr)

corMat <- round(cor(x=hs.per.Sample, y=Interop.per.Sample, method = "spearman", use = "complete.obs"), 2)



hs.cor.heatmap <- ggplot(data = melt(corMat, na.rm = T), aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile(colour = "white") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                       midpoint = 0, limit = c(-1,1), space = "Lab", 
                       name="Spearman\nCorrelation") + 
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                   size = 12, hjust = 1)) + xlab("Hs Metrics") + ylab("Interop") + ggtitle("Interop vs hs metrics")

hs.cor.heatmap
 ggsave(
  "hs.png",
  plot = hs.cor.heatmap,
  dpi = 300,
  limitsize = T)
```
 

```{r}
library(ggplot2)
library(reshape2)
library(dplyr)

corMat <- round(cor(x=fl.per.Sample, y=Interop.per.Sample, method = "spearman", use = "complete.obs"), 2)



fl.cor.heatmap <- ggplot(data = melt(corMat, na.rm = T), aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile(colour = "white") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                       midpoint = 0, limit = c(-1,1), space = "Lab", 
                       name="Spearman\nCorrelation") + 
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                   size = 12, hjust = 1)) + xlab("Hs Metrics") + ylab("Interop") + ggtitle("Correlations between Interop vs hs metrics")

fl.cor.heatmap
```

```{r}
library(DBI)
library(knitr)
library(ggplot2)

## establish connection to sqlite db
connection = dbConnect(RSQLite::SQLite(), "data/SQLITE/databaseFull.db")

res <- dbSendQuery(connection, "
                   Select 
                   avg(QualityByCycle.Quality) as average_cycle_quality,
                  avg(RunSummary.Yield) as Yield,
                  avg(RunSummary.ProjectedYield) as ProjectedYield,
                  avg(RunSummary.Intensity) as Intensity,
                  avg(RunSummary.Q30) as RunQ30,
                  avg(Lanes.Tiles) as Tiles,
                  avg(Lanes.DensityMIN) as DensitySD,
                  avg(Lanes.DensityMAX) as ClusterDensity,
                  avg(Lanes.ClusterMIN) as ClusterSD,
                  avg(Lanes.ClusterMAX) as ClustersPCT,
                  avg(Lanes.LegacyPhasing) as LegacyPhasing,
                  avg(Lanes.LegacyPrePhasing) as LegacyPrePhasing,
                  avg(Lanes.PhasingSlope) as PhasingSlope,
                  avg(Lanes.PrePhasingSlope) as PrePhasingSlope,
                  avg(Lanes.PhasingOffset) as PhasingOffset,
                  avg(Lanes.PrePhasingOffset) as PrePhasingOffset From Samples 
                  INNER JOIN RUNS ON Samples.Sequencer==RUNS.Sequencer AND Samples.Run==RUNS.Number AND Samples.startDate==RUNS.Date 
                  INNER JOIN RunSummary ON RUNS.UniqueID==RunSummary.UniqueID
                  INNER JOIN Lanes ON RUNS.UniqueID==Lanes.UniqueID
                  INNER JOIN QualityByCycle ON Samples.ID==QualityByCycle.SampleID
                  GROUP BY Samples.ID
                  ORDER BY RUNS.Date")
Avg.cycle.Qual.per.Sample <- dbFetch(res)
dbClearResult(res)
rm(res)
```



* PFmismatchRate Vs Intensity
* PFmismatchRate Vs RunQ30
* PFmismatchRate Vs ClusterDensity
* PFmismatchRate Vs ClusterPCT
* PFmismatchRate Vs Phasing
* PFmismatchRate Vs PrePhasing

* PFHQErrorRate Vs Intensity
* PFHQErrorRate Vs RunQ30
* PFHQErrorRate Vs ClusterDensity
* PFHQErrorRate Vs ClusterPCT
* PFHQErrorRate Vs PhasingOffset

* IndelRate (Insertions and deletions per 100 aligned bases) vs Read Lengts/Yield
* AdapterPercentage (The fraction of PF reads that are unaligned and match to a known adapter sequence right from the start of the read) vs Phasing offset

The ErrorRate and Mismatch Rate seem to be identical in terms of correlation..

```{r}
library("ggpubr")
library(gridExtra)
p1 <- ggscatter(Metrics.per.Sample, x = "PFHQErrorRate", y = "PFmismatchRate", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson",
          xlab = "HQ Error Rate", ylab = "Mismatch Rate")

p1
```
Infact the Error Rate and Mismatch Rate are identical, with a p-value of 2.2e-16.

PF_MISMATCH_RATE: the rate of bases mismatching the refference for all bases aligned to the reference sequence
PF_HQ_ERROR_RATE: The fraction of bases that mismatch the reference in PF HQ aligned reads

I suspect that if all aligned bases passed the filter, you get the same values here. We can say the same for the total reads/reads passing filter. 

```{r}
library(ggplot2)
library(reshape2)
library(dplyr)

corMat <- round(cor(y=Avg.cycle.Qual.per.Sample[1], x=Avg.cycle.Qual.per.Sample[2:15], method = "spearman", use = "complete.obs"), 2)



Q.cor.heatmap <- ggplot(data = melt(corMat, na.rm = T), aes(x=Var1, y=Var2, fill=value)) + 
  geom_tile(colour = "white") +
  scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                       midpoint = 0, limit = c(-1,1), space = "Lab", 
                       name="Spearman\nCorrelation") + 
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                   size = 12, hjust = 1)) + xlab("Hs Metrics") + ylab("Interop") + ggtitle("Correlations between Interop vs hs metrics")

Q.cor.heatmap
```


```{r}
library(igraph)
num <- subset(Metrics.per.Sample, select = -c(Weekday, Sequencer, Kit, BadCycles, BaitDesignEfficientcy, SupplementaryPass))
corMat <- cor(num, method = "spearman", use = "complete.obs")
corMat[corMat < 0.995] <- 0
network <- graph_from_adjacency_matrix(corMat, weighted=T, mode="undirected", diag=F)
plot(network)
```

```{r}
library(apcluster)
corMat <- round(cor(y=Interop.per.Sample, x=dependend, method = "spearman", use = "complete.obs"), 2)

apclust.metrics <- apcluster(negDistMat(r = 2), corMat)

apclust.metrics
cat("affinity propogation optimal number of clusters:", length(apclust.metrics@clusters), "\n")
```
```{r}
library(apcluster)
corMat <- round(cor(y=dependend, x=Interop.per.Sample, method = "spearman", use = "complete.obs"), 2)

apclust.metrics <- apcluster(negDistMat(r = 2), corMat)

apclust.metrics
cat("affinity propogation optimal number of clusters:", length(apclust.metrics@clusters), "\n")
```


